### ISSUE TITLE
Move MItem Category to front of ItemDiscussion.

### ISSUE THREAD

**Issue by Megalomaniak** [TwistedTwigleg/vanilla-bootstrap#15]
So this:
![DiscussionItem-old](https://user-images.githubusercontent.com/3162126/55327070-ef1a9a80-5491-11e9-8c3c-34440d70c368.png)
becomes something like this:
![DiscussionItem-new-mockup](https://user-images.githubusercontent.com/3162126/55327079-f3df4e80-5491-11e9-951e-ad805f6c69c9.png)
<sub>_Note that the latter is just a quick mockup._</sub>
- [x] Re-position item
- [x] Resize item
- [x] _**Bonus task:** look into implementing unique colors per category!_
- [x] Implement for other pages listing discussions from a variety of categories.

**Comment by TwistedTwigleg** [comment]
Makes it much easier to see what goes where at a glance. I really like this idea, I think it would definitely improve the usability of the forums.

**Comment by TwistedTwigleg** [comment]
I think I have sorta got this working.

I was going to use CSS, but because the meta data where the MItem Category is located under is after the profile picture, there didn't appear to be a way to move it using just CSS.

It needs a bit of work, but with some custom Javascript I have found a way to add the category link before the profile picture.
Really all that I'm missing right now is the CSS to style it. I also need to hide or remove the MItem Category, and I probably should clean up the Javascript code since it is kinda a mess.

I'll try to work more on this soon and commit the changes.

**Comment by TwistedTwigleg** [comment]
I know this issue is closed, but I already had the majority of the work done so I pushed a commit that makes these changes.

I figure if we like it, then we can keep it, and if we don't then it is easy enough to remove, we just amend/remove the single commit that adds the functionality. No biggie either way.

I am not 100% sure on the styling and colors, but I think it will serve as a good starting point if we want to have the MItem Category links this way.

**Comment by Megalomaniak** [comment]
Yeah, I closed it since i didn't want to hack it with js and a css solution, while possible would have been even more hackish and potentially ruined the user experience of those that might use screen readers(the blind for an example).

The proper solution is a php plugin for vanilla, or implement the functionality within the theme addon via themehooks however that makes it complex enough I'd rather push it into a future consideration.

Note that I closed it in a tab that had not been refreshed in a while so I had not noticed your reply here.

**Comment by TwistedTwigleg** [comment]
> Yeah, I closed it since i didn't want to hack it with js and a css solution, while possible would have been even more hackish and potentially ruined the user experience of those that might use screen readers(the blind for an example).
> The proper solution is a php plugin for vanilla, or implement the functionality within the theme addon via themehooks however that makes it complex enough I'd rather push it into a future consideration.

Fair enough. I had not considered users with things like screen readers. I can see how having custom HTML injected could cause issues in those cases.

I removed the code from the theme and slapped it into a plugin instead. Because it is a plugin separate from the theme, I was able to make it where you can turn the functionality the plugin provides on and off on a per user basis, akin to the theme chooser plugin.

That way, people who want to leave the categories the way there are and/or have issues with custom injected HTML (like screen readers) can turn off the plugin's functionality from their profile page.

> Note that I closed it in a tab that had not been refreshed in a while so I had not noticed your reply here.

No worries.

____

I updated the test forum with the latest changes, including the new plugin, so now we just have to wait and see what others think 🙂

**Comment by Megalomaniak** [comment]
Care to share the plugin, preferably as a  git repo?

edit: also whats the plugin you used for theme selection?

**Comment by TwistedTwigleg** [comment]
> Care to share the plugin, preferably as a git repo?

Sure, I can make a repository for it. I'll see about getting something up soon-ish. I want to make a few changes before making a repository for the plugin. There are a few things I want to change code-wise, and I want to write out some comments/documentation.

> edit: also whats the plugin you used for theme selection?

I'm using a slightly modified version of the [themechooser-plugin](https://open.vanillaforums.com/addon/themechooser-plugin) on the Vanilla addons site. If I recall correctly, the changes I made were changing the path to it gets the CSS from `themes/bootstrap`, and I made it remove the original CSS before adding a new one.

That said, I don't totally remember what modifications I made to get it working. I'd have to double check. Likewise, I'll see about posting the changes I made to a repository or something.

_____

Until I have a repository up and going, I have attached a zip file containing all of the plugins I've edited one way or another. This includes `ThemeChooser`, `TagsExtended`, `ImgurUpload` (sorta working), and `DiscussionCategoryStyleExtended`.

The changes are fairly minor, with the exception of `DiscussionCategoryStyleExtended`, which I created from scratch. [Edited_Plugins_04_03_2019.zip](https://github.com/TwistedTwigleg/vanilla-bootstrap/files/3041229/Edited_Plugins_04_03_2019.zip)

**Comment by Megalomaniak** [comment]
Great, thanks.

**Comment by Megalomaniak** [comment]
_Righty-O_, so I have refactored the plugin to add a div in there, and tweaked the css so it all now looks closer to my mockup:

<details>
<summary>Image</summary>

![CategoryLable_20190404_Mega](https://user-images.githubusercontent.com/3162126/55581875-a92a3480-5726-11e9-8cab-096dcb906691.png)
</details>

However as you can see there's a few issues with the way our theme is rendering the rest of the contents of the list item. None of that is a bug with the plugin though, the theme needs adjustments now and then it should look pretty much perfect(only font size would need adjustment), or as close as it can, anyways.

Another thing worth pointing out is that it currently only works on the discussions page, however this should probably apply also in other places, such as the bookmarks page. Since the current plugin can probably be considered a prototype for testing, I think it's fine though. Get it working in one place first, then consider expanding it to other places.

Best of course would be to convert the whole logic into php and perhaps even have the php generate the css from the categories automatically however that's probably best left into the future.

Can't seem to upload the .zip in here so I've created a gist for them:
https://gist.github.com/Megalomaniak/7559d5f3c8f14988d0f4e2930646c2eb

**Comment by Megalomaniak** [comment]
As you can see from the image in previous post the issues with the theme now are:
- [ ] Fix: Topic title breaks into new line causing the list item to become taller.
- [ ] Fix: Topic metadata/details and tags are too lock breaking into new line...

Fix for topic title is simple, if it's too long: truncate and add ellipsis to the end. Should probably also make hover show the full title then as alt-text.

The other one is a bit more complex and has multiple potential fixes, will leave that for later discussion.

**Comment by TwistedTwigleg** [comment]
Cool! I’ll have to take a closer look at everything once I’m able. My schedule is such that I won’t really have much spare time today, unfortunately.

> Another thing worth pointing out is that it currently only works on the discussions page, however this should probably apply also in other places, such as the bookmarks page. Since the current plugin can probably be considered a prototype for testing, I think it's fine though. Get it working in one place first, then consider expanding it to other places.

Adding it to other pages should be easy enough. We just need to use the right function names in the PHP file so we can add the CSS and Javascript files. I can look into this and see about making the changes needed.

> Best of course would be to convert the whole logic into php and perhaps even have the php generate the css from the categories automatically however that's probably best left into the future.

I suppose, though personally I’m fine with the Javascript solution. I don’t see what advantages we’d gain using PHP, other than I suppose the code would be in PHP and not Javascript, but that is more a style/programming-language-preference thing than anything else.

Before I wrote the plugin, I looked into injecting HTML code into a document with PHP. As far as I could find you can only inject HTML into either the head of the document of the foot. There didn’t appear to be any way to inject HTML into a specific part of a page in PHP. That said, my PHP skills are only ‘meh’, and I felt more comfortable using Javascript anyway so I just used that.

That said, if you want to convert it to PHP, by all means. I just personally don’t see any reason to rewrite it to use PHP.

> As you can see from the image in previous post the issues with the theme now are:
> 
> * Topic title breaks into new line causing the list item to become taller.
> * Topic metadata/details and tags are too lock breaking into new line...
> 
> Fix for topic title is simple, if it's too long: truncate and add ellipsis to the end. Should probably also make hover show the full title then as alt-text.

That works. It should just be a matter of adding some CSS.

Or we could just add some padding and leave it as is. Granted it knocks everything in the meta data further down, but that might not necessarily be a bad thing. If we just knock the meta data down by default, that would also make it easier to solve the second issue, since there would be more room for the meta data (so more rooms for things like tags).

Something to think about I suppose.

> The other one is a bit more complex and has multiple potential fixes, will leave that for later discussion.

Sounds good to me.

____

Thanks for making these changes! I’ll take a closer look once I can.

**Comment by Megalomaniak** [comment]
> Cool! I’ll have to take a closer look at everything once I’m able. My schedule is such that I won’t really have much spare time today, unfortunately.

Yeah, no problem. We get to things when we can. Tomorrow is going to likely be very busy for me.

**Comment by TwistedTwigleg** [comment]
Okay, I made a [new repository](https://github.com/TwistedTwigleg/godot-forums-vanilla-plugins) so we can work on plugins. Right now it only includes the plugins that we've modified, or made custom for the forums.

I also pulled in the changes you made to the category discussion style extended plugin to the repository. 

I'm thinking it may be best to wait before updating the plugin on the test forums until we have fixed the two issues above. That said, if you want I'll update the plugin on the test forums.

**Comment by Megalomaniak** [comment]
Hmm, yeah, that's what I was originally thinking too, but maybe it will spur us on to fix 'em more quickly. The above issues are with the theme itself, not really the plugin after all. Not sure. I'll leave it up to you.

**Comment by TwistedTwigleg** [comment]
> Hmm, yeah, that's what I was originally thinking too, but maybe it will spur us on to fix 'em more quickly. The above issues are with the theme itself, not really the plugin after all. Not sure. I'll leave it up to you.

I think I'll leave the plugin on the test server alone for now then, since it is mainly visual and leaving it shouldn't hurt anything.

That said, I'm testing an idea of using a flex layout with the plugin, and looking at it we might want to use it in the theme just in general. It *might* make things easier in regards to both issues above.

Right now I'm using the CSS injected by the plugin for quick comparisons. I'll push a change to the plugin repository in a little bit. If we like it, then we can incorporate the flex layout into the main theme.

Edit: Commit pushed! Now the category label container is always the full height of the list item. I'm not 100% sure if using the flex stuff is good or not... More testing is probably needed.

**Comment by Megalomaniak** [comment]
I've been busy for the last few days but I'll try to take a look at it either today or tomorrow latest.

**Comment by TwistedTwigleg** [comment]
No worries, take as long as you need! I’m in no rush.

**Comment by TwistedTwigleg** [comment]
I made some changes to the GodotDarkly theme and updated the theme on the test server. The changes are minor, mainly just copying things over from GodotLightly and making small edits here and there as needed.

However, because I was already uploading the latest theme updates, I decided to also upload the changes I made to the discussion category plugin and the Imgur plugin to the test forum as well. It looks like the discussion category plugin is doing okay, so fingers crossed that it will stay that way.
